### PR TITLE
QR codes to PNG

### PR DESCRIPTION
**Pull request by FrankAbagnaleJr**
Saves QR codes to PNG in a subdirectory ("qr") of the current directory, rather than attempting to display them in the console.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -21,7 +21,7 @@ The protobuf package of Google for proto3 is required for running this script.
 
 For printing QR codes, the qrcode module is required
 
-    pip install qrcode
+    pip install qrcode[pil]
```
Comment (scito) [inline README.md:24]: What is `[pil]` for?

File: extract_otp_secret_keys.py [hunk 2]
```diff
@@ -48,6 +48,7 @@
 from urllib.parse import parse_qs, urlencode, urlparse, quote
 
 import generated_python.google_auth_pb2
+import os,re,string
```
Comment (scito) [inline extract_otp_secret_keys.py:51]: I suggest using from ... import

File: extract_otp_secret_keys.py [hunk 3]
```diff
@@ -48,6 +48,7 @@
...
 
 arg_parser = argparse.ArgumentParser()
 arg_parser.add_argument('--verbose', '-v', help='verbose output', action='store_true')
```
Comment (scito) [inline extract_otp_secret_keys.py:54]: I suggest using an option to support both printing to tty and saving to file. Something like
`arg_parser.add_argument('--save, '-s', help='save QR codes to subdirectory qr', action='store_true')`

And separate printing and saving in the code below.

File: extract_otp_secret_keys.py [hunk 4]
```diff
@@ -85,7 +88,9 @@ def print_qr(data):
     if verbose: print(payload)
 
     # pylint: disable=no-member
+    i = int(0)
     for otp in payload.otp_parameters:
+        i+=1
```
Comment (scito) [inline extract_otp_secret_keys.py:93]: Code formatting: I've used spaces around operations in this project.
I suggest
`i += 1`
Here and also in other parts of this pull request

File: extract_otp_secret_keys.py [hunk 5]
```diff
@@ -97,4 +102,9 @@ def print_qr(data):
         otp_url = 'otpauth://{}/{}?'.format('totp' if otp.type == 2 else 'hotp', quote(otp.name)) + urlencode(url_params)
         if args.qr:
             if verbose: print(otp_url)
-            print_qr(otp_url)
+            if not(os.path.exists("qr")):os.mkdir("qr")
```
Comment (scito) [inline extract_otp_secret_keys.py:105]: Code formatting:
I suggest either a space after : or put it on a new line

File: extract_otp_secret_keys.py [hunk 6]
```diff
@@ -85,7 +88,9 @@ def print_qr(data):
     if verbose: print(payload)
 
     # pylint: disable=no-member
+    i = int(0)
```
Comment (scito) [inline extract_otp_secret_keys.py:91]: Is int() necessary?

File: extract_otp_secret_keys.py [hunk 7]
```diff
@@ -97,4 +102,9 @@ def print_qr(data):
         otp_url = 'otpauth://{}/{}?'.format('totp' if otp.type == 2 else 'hotp', quote(otp.name)) + urlencode(url_params)
         if args.qr:
             if verbose: print(otp_url)
-            print_qr(otp_url)
+            if not(os.path.exists("qr")):os.mkdir("qr")
+            pattern = re.compile('[\W_]+')
+            file_otp_name = pattern.sub('', otp.name)
+            file_otp_issuer = pattern.sub('',otp.issuer)
```
Comment (scito) [inline extract_otp_secret_keys.py:108]: Formatting: I suggest adding a space
`file_otp_issuer = pattern.sub('', otp.issuer)`

**Review by scito — CHANGES_REQUESTED** [review]
Thanks for your proposal

File: README.md (hunk 1 above)
Comment (FrankAbagnaleJr) [reply README.md:24]: Installs the Pillow image library along with qrcode library to generate the image.
https://pypi.org/project/Pillow/

File: extract_otp_secret_keys.py (hunk 6 above)
Comment (FrankAbagnaleJr) [reply extract_otp_secret_keys.py]: It's not, I don't remember why I put that there. Taken out now.

File: extract_otp_secret_keys.py (hunk 7 above)
Comment (FrankAbagnaleJr) [reply extract_otp_secret_keys.py]: Done

File: extract_otp_secret_keys.py (hunk 5 above)
Comment (FrankAbagnaleJr) [reply extract_otp_secret_keys.py]: Done

File: extract_otp_secret_keys.py (hunk 3 above)
Comment (FrankAbagnaleJr) [reply extract_otp_secret_keys.py:54]: Done

File: extract_otp_secret_keys.py [hunk 8]
```diff
@@ -46,18 +46,27 @@
...
 
 verbose = args.verbose
+saveqr = args.saveqr
+printqr = args.printqr
+
 if args.qr:
     from qrcode import QRCode
+    if not(saveqr or printqr):
+        print("Error: If you specify to generate a QR code, you must specify how you want to receive it. Please specify -s or -p. Use -h or --help for more information")
+        exit()
```
Comment (scito) [inline extract_otp_secret_keys.py:69]: I think this part is not necessary.

File: extract_otp_secret_keys.py [hunk 9]
```diff
@@ -46,18 +46,27 @@
...
+arg_parser.add_argument('--saveqr', '-s', help='save QR code(s) as images to the "qr" subfolder', action='store_true')
+arg_parser.add_argument('--printqr', '-p', help='print QR code(s) as text to the terminal', action='store_true')
 arg_parser.add_argument('infile', help='file or - for stdin (default: -) with "otpauth-migration://..." URLs separated by newlines, lines starting with # are ignored')
 args = arg_parser.parse_args()
 
 verbose = args.verbose
+saveqr = args.saveqr
+printqr = args.printqr
+
 if args.qr:
```
Comment (scito) [inline extract_otp_secret_keys.py:65]: I propose
`if saveqr or printqr:`

File: extract_otp_secret_keys.py [hunk 10]
```diff
@@ -46,18 +46,27 @@
...
 from urllib.parse import parse_qs, urlencode, urlparse, quote
-
+from os import path, mkdir
+from re import sub, compile as rcompile
 import generated_python.google_auth_pb2
 
 arg_parser = argparse.ArgumentParser()
 arg_parser.add_argument('--verbose', '-v', help='verbose output', action='store_true')
-arg_parser.add_argument('--qr', '-q', help='print QR codes (otpauth://...)', action='store_true')
+arg_parser.add_argument('--qr', '-q', help='generate QR codes (otpauth://...)', action='store_true')
```
Comment (scito) [inline extract_otp_secret_keys.py:55]: I suggest to remove --qr parameter and just use --saveqr and --printqr

**Review by scito — CHANGES_REQUESTED** [review]
Thanks for contribution!
I really appreciate it.
I think after removal of --qr parameter. I'll merge this pull request.

File: extract_otp_secret_keys.py (hunk 8 above)
Comment (FrankAbagnaleJr) [reply extract_otp_secret_keys.py]: removed

File: extract_otp_secret_keys.py (hunk 9 above)
Comment (FrankAbagnaleJr) [reply extract_otp_secret_keys.py]: done

File: extract_otp_secret_keys.py (hunk 10 above)
Comment (FrankAbagnaleJr) [reply extract_otp_secret_keys.py]: done

**Pull request merged**